runtime/value.go

```diff
@@ -0,0 +1,355 @@
+	Key() string
+
+	// Start starts listening to value updates.
+	Start() error
+
+	// Update updates the value with a kv value.
+	Update(value kv.Value) error
+
+	// Stop stops listening to value updates.
+	Stop() error
+
+	// Close closes the value.
+	Close() error
+}
+
+type value struct {
+	sync.RWMutex
+
+	key         string
+	store       kv.Store
+	opts        Options
+	log         xlog.Logger
+	nowFn       clock.NowFn
+	unmarshalFn UnmarshalFn
+	versionFn   VersionFn
+	cutoverFn   CutoverFn
+	processFn   ProcessFn
+	updateFn    updateFn
+
+	watch       kv.ValueWatch
+	initialized bool
+	status      valueStatus
+	version     int
+	updateWg    sync.WaitGroup
+	closedCh    chan struct{}
+}
+
+// NewValue creates a new value.
+func NewValue(
+	key string,
+	opts Options,
+) Value {
+	v := &value{
+		key:         key,
+		opts:        opts,
+		store:       opts.KVStore(),
+		log:         opts.InstrumentOptions().Logger(),
+		nowFn:       opts.ClockOptions().NowFn(),
+		unmarshalFn: opts.UnmarshalFn(),
+		versionFn:   opts.VersionFn(),
+		cutoverFn:   opts.CutoverFn(),
+		processFn:   opts.ProcessFn(),
+		initialized: false,
+		status:      valueStopped,
+		version:     defaultVersion,
+		closedCh:    make(chan struct{}),
+	}
+	v.updateFn = v.Update
+	return v
+}
+
+func (v *value) Key() string { return v.key }
+
+func (v *value) Start() error {
+	v.Lock()
+	if v.status != valueStopped {
+		v.Unlock()
+		return errValueNotStopped
+	}
+	watch, err := v.store.Watch(v.key)
+	if err != nil {
+		v.Unlock()
+		return err
+	}
+	v.watch = watch
+	v.status = valueStarted
+	v.Unlock()
+
+	select {
+	case <-watch.C():
+	case <-time.After(v.opts.InitWatchTimeout()):
+		err = errInitWatchTimeout
+	}
+
+	if err == nil {
+		err = v.updateFn(watch.Get())
+	}
+
+	// NB(xichen): we want to start watching updates even though
+	// we may fail to initialize the value temporarily (e.g., during
+	// a network partition) so the value will be updated when the
+	// error condition is resolved.
+	v.updateWg.Add(1)
+	go v.watchUpdates(watch)
+
+	return err
+}
+
+func (v *value) Update(value kv.Value) error {
+	v.RLock()
+	if v.status == valueClosed {
+		v.RUnlock()
+		return errValueClosed
+	}
+	initialized := v.initialized
+	v.RUnlock()
+
+	if !initialized {
+		return v.initUpdate(value)
+	}
+	return v.cachedUpdate(value)
+}
+
+func (v *value) Stop() error {
+	v.Lock()
+	if v.status != valueStarted {
+		v.Unlock()
+		return errValueNotStarted
+	}
+	if v.watch != nil {
+		v.watch.Close()
+	}
+	v.watch = nil
+	v.status = valueStopping
+	v.Unlock()
+
+	// NB(xichen): we need to release lock while waiting or
+	// otherwise the stopping goroutine and the updating goroutine
+	// may end up in a deadlock situation.
+	v.updateWg.Wait()
+
+	v.Lock()
+	defer v.Unlock()
+
+	// The value may be closed at this point, in which case
+	// we don't override its status.
+	if v.status == valueStopping {
+		v.status = valueStopped
+	}
+	return nil
+}
+
+func (v *value) Close() error {
```

Comment:
I see, might it be worth renaming `Start` and `Stop` to `Watch` and `Unwatch` perhaps to make it a little more clear on the semantics of start/stopping is?